rovercode/app.py

```diff
@@ -0,0 +1,181 @@
+        }
+        for sensor in binary_sensors:
+            time.sleep(0.2)
+            sensor_message[constants.SENSOR_ID_FIELD] = sensor.name
+            changed_value = sensor.get_change()
+            if changed_value is not None:
+                sensor_message[constants.SENSOR_VALUE_FIELD] = changed_value
+                ws.send(json.dumps(sensor_message))
+        if run_once_only:
+            break
+        time.sleep(0.2)  # pragma: no cover
+
+
+def on_message(ws, raw_message):
+    """Handle incoming websocket message."""
+    global motor_controller
+    message = json.loads(raw_message)
+    type = message[constants.MESSAGE_TYPE]
+    if type == constants.MOTOR_COMMAND:
+        if (message[constants.MOTOR_ID_FIELD] not in constants.MOTOR_IDS):
+            LOGGER.warning(
+                "Invalid motor {}".format(message[constants.MOTOR_ID_FIELD]))
+            return
+        motor_controller.set_speed(message[constants.MOTOR_ID_FIELD],
+                                   message[constants.MOTOR_VALUE_FIELD])
+
+
+def on_error(ws, error):  # pragma: no cover
+    """Handle error from websocket connection."""
+    LOGGER.error(error)
+
+
+def on_close(ws):  # pragma: no cover
+    """Handle closing of websocket connection."""
+    LOGGER.warning("Websocket connection closed")
+
+
+def on_open(ws):
+    """Start up threads upon opening websocket connections."""
+    global session
+    global motor_controller
+    global rovercode_web_url
+    global client_id
+
+    info = json.loads(session.get('{}/api/v1/rovers?client_id={}'.format(rovercode_web_url, client_id)).text)[0]
+    LOGGER.info("Found myself - I am %s, with id %s", info[constants.ROVER_NAME], info[constants.ROVER_ID])
+
+    # Create motor manager
+    motor_controller = MotorController(info['left_forward_pin'], info['left_backward_pin'],
+                                       info['right_forward_pin'], info['right_backward_pin'],
+                                       AdafruitPwmManager())
+
+    # Start heartbeat thread
+    t = Thread(target=send_heartbeat, args=(ws,))
+    t.start()
+    LOGGER.info("Heartbeat thread started")
+
+    # Start inputs thread
+    binary_sensors = init_inputs(info, dummy='rovercode.com' not in rovercode_web_url)
+    t = Thread(target=poll_sensors, args=(ws, binary_sensors))
+    t.start()
+    LOGGER.info("Sensors thread started")
+
+
+def run_service(run_forever=True, use_dotenv=True):
+    """Kick off the service."""
+    global session
+    global motor_controller
+    global rovercode_web_url
+    global client_id
+    LOGGER.info("Starting the Rovercode service!")
+    if use_dotenv:  # pragma: no cover
+        load_dotenv('../.env')
+    rovercode_web_host = os.getenv("ROVERCODE_WEB_HOST", "rovercode.com")
+
+    rovercode_web_host_secure = os.getenv("ROVERCODE_WEB_HOST_SECURE", 'True').lower() == 'true'
+    if rovercode_web_host[-1:] == '/':  # pragma: no cover
```

Comment:
Why can't these cases be covered?

Reply:
No good reason. Care if I write a ticket for it, though?